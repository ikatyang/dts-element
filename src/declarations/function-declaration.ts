import * as ts from 'typescript';
import { ElementKind } from '../constants';
import { create_element, IElement, IElementOptions } from '../element';
import { transform } from '../transform';
import { create_function_type, IFunctionType } from '../types/function-type';
import { add_declare_modifier_if_need } from '../utils';

export interface IFunctionDeclarationOptions extends IElementOptions {
  name: string | undefined;
  type?: IFunctionType;
  export?: boolean;
}

export interface IFunctionDeclaration
  extends IElement<ElementKind.FunctionDeclaration>,
    IFunctionDeclarationOptions {}

export const create_function_declaration = (
  options: IFunctionDeclarationOptions,
): IFunctionDeclaration => ({
  ...create_element(ElementKind.FunctionDeclaration),
  ...options,
});

// tslint:disable:ter-indent

/**
 * @hidden
 */
export const transform_function_declaration = (
  element: IFunctionDeclaration,
  path: IElement<any>[],
) => {
  const function_type = transform(
    element.type || create_function_type(),
    path,
  ) as ts.FunctionDeclaration;
  return ts.createFunctionDeclaration(
    /* decorators      */ undefined,
    /* modifiers       */ add_declare_modifier_if_need(
      element.export === true
        ? [ts.createToken(ts.SyntaxKind.ExportKeyword)]
        : undefined,
      path,
    ),
    /* asteriskToken   */ undefined,
    /* name            */ element.name,
    /* typeParameters  */ function_type.typeParameters,
    /* parameters      */ function_type.parameters,
    /* type            */ function_type.type,
    /* body            */ undefined,
  );
};
