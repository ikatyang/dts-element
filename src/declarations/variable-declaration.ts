import * as ts from 'typescript';
import { IType } from '../collections';
import { any_type, ElementKind } from '../constants';
import {
  create_element,
  is_element,
  IElement,
  IElementOptions,
} from '../element';
import { transform } from '../transform';
import { add_declare_modifier_if_need } from '../utils';

export interface IVariableDeclarationOptions extends IElementOptions {
  name: string;
  type?: IType;
  export?: boolean;
  const?: boolean;
  let?: boolean;
}

export interface IVariableDeclaration
  extends IElement<ElementKind.VariableDeclaration>,
    IVariableDeclarationOptions {}

export const create_variable_declaration = (
  options: IVariableDeclarationOptions,
): IVariableDeclaration => ({
  ...create_element(ElementKind.VariableDeclaration),
  ...options,
});

export const is_variable_declaration = (
  value: any,
): value is IVariableDeclaration =>
  is_element(value) && value.kind === ElementKind.VariableDeclaration;

// tslint:disable:ter-indent

/**
 * @hidden
 */
export const transform_variable_declaration = (
  element: IVariableDeclaration,
  path: IElement<any>[],
) => {
  const type = transform(element.type || any_type, path) as ts.TypeNode;
  return ts.createVariableStatement(
    /* modifiers       */ add_declare_modifier_if_need(
      element.export === true
        ? [ts.createToken(ts.SyntaxKind.ExportKeyword)]
        : undefined,
      path,
    ),
    /* declarationList */ ts.createVariableDeclarationList(
      /* declarations  */ [ts.createVariableDeclaration(element.name, type)],
      /* flags         */ element.const === true
        ? ts.NodeFlags.Const
        : element.let === true ? ts.NodeFlags.Let : undefined,
    ),
  );
};
