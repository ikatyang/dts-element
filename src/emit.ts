import * as ts from 'typescript';
import { IElement } from './element';
import { transform } from './transform';

export const emit_native = (node: ts.Node, options?: ts.PrinterOptions) => {
  const source_file = ts.createSourceFile(
    '',
    '',
    ts.ScriptTarget.Latest,
    false,
    ts.ScriptKind.TS,
  );
  const printer = ts.createPrinter(options);
  return printer.printNode(ts.EmitHint.Unspecified, node, source_file);
};

export const emit = (element: IElement<any>, options?: ts.PrinterOptions) =>
  emit_native(transform(element), options);
