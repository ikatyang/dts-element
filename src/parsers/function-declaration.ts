import * as ts from 'typescript';
import {
  create_function_declaration,
  IFunctionDeclaration,
} from '../declarations/function-declaration';
import { IGenericDeclaration } from '../declarations/generic-declaration';
import { IParameterDeclaration } from '../declarations/parameter-declaration';
import { parse_native } from '../parse';
import { create_function_type } from '../types/function-type';
import { if_defined } from '../utils';
import { parse_identifier } from './identifier';

export const parse_function_declaration = (
  node: ts.FunctionDeclaration,
): IFunctionDeclaration =>
  create_function_declaration({
    name: parse_identifier(node.name as ts.Identifier),
    type: create_function_type({
      generics: if_defined(
        node.typeParameters,
        generics => generics.map(parse_native) as IGenericDeclaration[],
      ),
      parameters: if_defined(
        node.parameters,
        parameters => parameters.map(parse_native) as IParameterDeclaration[],
      ),
      return: if_defined(node.type, parse_native),
    }),
  });
