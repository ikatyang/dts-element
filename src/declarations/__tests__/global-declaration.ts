import { emit } from '../../emit';
import { create_function_declaration } from '../function-declaration';
import {
  create_global_declaration,
  is_global_declaration,
} from '../global-declaration';
import { create_variable_declaration } from '../variable-declaration';

it('should return correctly', () => {
  expect(emit(create_global_declaration())).toMatchSnapshot();
});

it('should return correctly with members', () => {
  expect(
    emit(
      create_global_declaration({
        members: [
          create_variable_declaration({
            name: 'a',
          }),
          create_function_declaration({
            name: 'b',
          }),
        ],
      }),
    ),
  ).toMatchSnapshot();
});

describe('is_global_declaration', () => {
  it('should return correctly', () => {
    const element = create_global_declaration();
    expect(is_global_declaration(element)).toBe(true);
  });
});
