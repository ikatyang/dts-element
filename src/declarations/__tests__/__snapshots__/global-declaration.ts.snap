// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return correctly 1`] = `"declare global { }"`;

exports[`should return correctly with members 1`] = `
"declare global {
    var a: any;
    function b(): any;
}"
`;
